core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -168,68 +170,70 @@ protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void cont
-            if (leftEstimate.isOutputRowCountUnknown()) {
-                smallestKnownEstimate = rightEstimate;
-            }
-            else if (rightEstimate.isOutputRowCountUnknown()) {
-                smallestKnownEstimate = leftEstimate;
-            }
-            else {
-                smallestKnownEstimate = leftEstimate.getOutputRowCount() <= rightEstimate.getOutputRowCount() ? leftEstimate : rightEstimate;
-            }
-            if (smallestKnownEstimate.isOutputRowCountUnknown()) {
+            Optional<PlanNodeStatsEstimate> smallest = terms.stream()
+                    .map(this::process)
+                    .filter(termEstimate -> !termEstimate.isOutputRowCountUnknown())
+                    .sorted(Comparator.comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
+                    .findFirst();
+
+            if (smallest.isEmpty()) {
                 return PlanNodeStatsEstimate.unknown();
             }
-            return smallestKnownEstimate.mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+
+            return smallest.get().mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
         }
 
-        private PlanNodeStatsEstimate estimateLogicalOr(Expression left, Expression right)
+        private PlanNodeStatsEstimate estimateLogicalOr(List<Expression> terms)
         {
-            PlanNodeStatsEstimate leftEstimate = process(left);
-            if (leftEstimate.isOutputRowCountUnknown()) {
-                return PlanNodeStatsEstimate.unknown();
+            PlanNodeStatsEstimate andEstimate = process(terms.get(0));
+            for (int i = 1; i < terms.size(); i++) {
+                andEstimate = new FilterExpressionStatsCalculatingVisitor(andEstimate, session, types).process(terms.get(i));
+
+                if (andEstimate.isOutputRowCountUnknown()) {
+                    return PlanNodeStatsEstimate.unknown();
+                }
             }
 
-            PlanNodeStatsEstimate rightEstimate = process(right);
-            if (rightEstimate.isOutputRowCountUnknown()) {
-                return PlanNodeStatsEstimate.unknown();
+            PlanNodeStatsEstimate summed = process(terms.get(0));
+            for (int i = 1; i < terms.size(); i++) {
+                summed = addStatsAndSumDistinctValues(summed, process(terms.get(i)));
             }
 
-            PlanNodeStatsEstimate andEstimate = new FilterExpressionStatsCalculatingVisitor(leftEstimate, session, types).process(right);
-            if (andEstimate.isOutputRowCountUnknown()) {
+            if (summed.isOutputRowCountUnknown()) {
                 return PlanNodeStatsEstimate.unknown();
             }
 
             return capStats(
-                    subtractSubsetStats(
-                            addStatsAndSumDistinctValues(leftEstimate, rightEstimate),
-                            andEstimate),
+                    subtractSubsetStats(summed, andEstimate),
```

Comment:
Indeed. The cardinality of the union of multiple sets is significantly more complicated. Let me figure out how it should work.

Reply:
This is the formula we need to implement here: https://en.wikipedia.org/wiki/Inclusion%E2%80%93exclusion_principle

Reply:
It can get expensive for large numbers of terms, as it needs to compute the intersections of all combinations of 3 sets, 4 sets, 5 sets ... up to the number of terms. We'll need an approximation.

Reply:
Yes, but I was thinking that this approach is too expensive and we're aiming for something cheaper. Maybe we should apply some arbitrary "overlap factor" if there are many operands.

Reply:
If I'm reading the old code correctly, it's computing the stats for `A OR B OR C OR D` with the equivalent of this formula:

```
| A OR B OR C OR D | = |A| + |B| + |C| + |D| - | C ∩ D | - | B ∩ C ∩ D | - | A ∩ B ∩ C ∩ D|
```

which is also very incorrect.

Reply:
@sopel39, do you remember what was the motivation for computing the stats for OR expressions that way?

Reply:
cc @losipiuk @kbajda

Reply:
> @sopel39, do you remember what was the motivation for computing the stats for OR expressions that way?

What are you asking for specifically?
This algorithm complies with `A + B - AB` (plus tweaks around capping and NDVs)
We assume we work on multiset here, hence we add `addStatsAndSumDistinctValues` and then `subtractSubsetStats` subtracts NDVs too

This is the original PR: https://github.com/prestodb/presto/pull/11973

Reply:
I restored the old logic, but in an iterative, non-recursive manner.